### PR TITLE
i#3170 accept custom [u]int64 definitions

### LINKED ISSUE #3170: Use C++11 definitions for [u]int64, when compiling on 64 bit
Currently, uint64 is defined as unsigned long int. That type is 64 bit if we compile with -m64, however, it's not the same type as unsigned long long: the former is "guaranteed to be at least 32 bits" (and happens to be 64 if -m64), the latter is 64 bits.

This causes linking problems in 64-bit projects linking a DR library exposing a function typed with uint64, if also the importing projects uses -DDR_DO_NOT_DEFINE_uint64: the importing project will compile the DR headers with their own (presumably, unsigned long long) uint64, and the mangled name of the import will differ from the mangled name of the definition.

### PR DESCRIPTION
**Pull request by mtrofin**
Allows specifying custom definitions for [u]int64, as well as formatting, if necessary. The macros are: DR_DEF_FOR_uint64, DR_DEF_FOR_int64, and INT64_FORMAT.

Fixes #3170

### CODE REVIEW

File: core/lib/globals_shared.h [hunk 1]
```diff
@@ -212,38 +212,52 @@ typedef byte *app_pc;
 typedef void (*generic_func_t)();
 
 #ifdef WINDOWS
-#    ifndef DR_DO_NOT_DEFINE_uint64
-typedef unsigned __int64 uint64;
+#    ifndef DR_DEF_FOR_uint64
```
Comment (derekbruening) [inline core/lib/globals_shared.h]: I think we have to turn off DR_DO_NOT_DEFINE if the corresponding DR_DEFINE_FOR is set.  The CHECK_TYPE_SIZE commands in our top-level CMakeLists.txt will auto-enable DR_DO_NOT_DEFINE, so an external project that's using CMake and has a type coming from a header checked by CHECK_TYPE_SIZE will not be able to use this DR_DO_NOT_DEFINE feature as-is.

[the remark this answers is not in the document]
Comment (mtrofin) [reply core/lib/globals_shared.h]: Done

File: suite/tests/client-interface/syscall.templatex [hunk 2]
```diff
@@ -13,7 +13,7 @@ stopping syscall monitoring
 starting syscall monitoring
 // I'd like to use "\r\n" but it turns into "\\r\\n" in CTestTestfile
 // and then doesn't match
-[0-9a-fx
+[0-9a-fx
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply suite/tests/client-interface/syscall.templatex]: done

File: core/lib/globals_shared.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply core/lib/globals_shared.h]: Ah, that's where it was. I looked under api/docs/release.dox. Should have grep-ed.

Comment (mtrofin) [reply core/lib/globals_shared.h]: done.

File: api/docs/intro.dox [hunk 3]
```diff
@@ -1922,6 +1922,8 @@ DynamoRIO, rebooting in safe mode will allow recovery.
  - The DynamoRIO header files have typedefs that may conflict with other
    header files wrapped in ifndef DR_DO_NOT_DEFINE_<type> to make it
    easier to work around such conflicts.
+ - Alternatively, use DR_DEFINE_FOR_<type>, in which case you may need to
```
Comment (derekbruening) [inline api/docs/intro.dox]: This doesn't make sense as a separate bullet item: "Alternatively" refers to the prior bullet, not anything else in this list.  Please merge into the prior one.

File: core/lib/globals_shared.h [hunk 4]
```diff
@@ -212,38 +212,60 @@ typedef byte *app_pc;
...
+#        endif
+#        ifndef INT64_FORMAT
+#            define INT64_FORMAT "ll"
 #        endif
-#        define INT64_FORMAT "ll"
 #    endif
 #endif
+
+#ifdef DR_DEFINE_FOR_uint64
+#undef DR_DO_NOT_DEFINE_uint64
```
Comment (derekbruening) [inline core/lib/globals_shared.h]: Looks like a clang-format violation -- prob why Travis failed.  Below too.

File: api/docs/intro.dox (hunk 3 above)
Comment (mtrofin) [reply api/docs/intro.dox]: Done

File: core/lib/globals_shared.h (hunk 4 above)
Comment (mtrofin) [reply core/lib/globals_shared.h]: Done. There was another reason, doxygen got confused by <type>_FORMAT. I fixed by enclosing in quotes, similar to what was done in the previous paragraph.

**Pull request merged**